src/docfx/restore/RestoreLock.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class RestoreLock
+    {
+        private readonly RestoreItem _restoreItem;
+
+        public RestoreLock(string docset)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            _restoreItem = Load(docset).Result;
+        }
+
+        public bool TryGetWorkTreeHead(string href, out string workTreeHead)
+        {
+            return _restoreItem.Git.TryGetValue(href, out workTreeHead);
+        }
+
+        public static Task Lock(string docset, Func<RestoreItem, RestoreItem> process)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            var restoreLockFilePath = GetRestoreLockFilePath(docset);
+            return ProcessUtility.ProcessLock(
+                Path.GetRelativePath(AppData.RestoreLockDir, restoreLockFilePath),
+                () =>
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(restoreLockFilePath));
+                    using (var fileStream = File.Open(restoreLockFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
```

Comment:
>fileStream [](start = 31, length = 10)

nit: It might be more readable to use

```
File.ReadAllText()
...
File.WriteAllText()
```

Reply:
I want to keep read and write in 1 IO

---
In reply to: [200884276](https://github.com/dotnet/docfx/pull/3020#discussion_r200884276) [](ancestors = 200884276)